cachi2/core/package_managers/yarn_classic/resolver.py

```diff
@@ -329,6 +353,36 @@ def resolve_packages(project: Project) -> Iterable[YarnClassicPackage]:
         chain(
             [_get_main_package(project.package_json)],
             _get_workspace_packages(project.source_dir, workspaces),
-            _get_packages_from_lockfile(project.source_dir, yarn_lock, runtime_deps),
+            _get_packages_from_lockfile(project.source_dir, mirror_dir, yarn_lock, runtime_deps),
         )
     )
+
+
+def _read_name_from_tarball(tarball_path: RootedPath) -> str:
+    """Read the package name from the package.json file in the cached tarball."""
+    with tarfile.open(tarball_path) as tar:
+        names = tar.getnames()
+        package_json_subpath = next((n for n in names if n.endswith("package.json")), None)
+        if package_json_subpath is None:
+            raise ValueError(f"No package.json found in the tarball {tarball_path}")
+
+        package_json = tar.extractfile(package_json_subpath)
+        if package_json is None:
+            raise ValueError(f"Failed to extract package.json from {tarball_path}")
+
+        package_json_content: dict[str, Any] = json.loads(package_json.read())
+
+        if (name := package_json_content.get("name")) is None:
+            raise ValueError(f"No 'name' field found in package.json in {tarball_path}")
+
+        return name
+
+
+def _read_name_from_package_json(path: RootedPath) -> str:
+    """Read the package name from a package.json file."""
+    package_json = PackageJson.from_file(path)
+
+    if (name := package_json.data.get("name")) is None:
```

Comment:
I wonder if this has to just move to `PackageJson` after- validator. Out of scope for this change though.